addons/hr_holidays/models/hr_leave.py

```diff
@@ -784,6 +784,17 @@ def _check_validity(self):
             for leave in leaves:
                 employees |= leave._get_employees_from_holiday_type()
             leave_data = leave_type.get_allocation_data(employees, date_from)
+
+            # Fetch the allocation records for validation
+            allocation_records = self.env['hr.leave.allocation'].search([
+                ('holiday_status_id', '=', leave_type.id),
+                ('employee_id', 'in', employees.ids),
+                ('state', '=', 'validate')
+            ])
+            # Validate against the allocation period
+            for allocation in allocation_records:
+                if leave.request_date_from < allocation.date_from or leave.request_date_to > allocation.date_to:
+                    raise ValidationError(_("The leave dates are outside the allocation period."))
```

Comment:
When the employee has an allocation from January 1st with end date at January 31st (only valid in January) and then a second allocation starting from February 2nd and no end date. Both are validated. The employee requests a time off for March 1st. That leave request is valid because the employee has an allocation that is valid for March.
As far as I can tell this code would throw a ValidationError because there is an allocation outside of the requested leave. Wouldn't it be better to add the start and end date to the domain and check if any allocation is loaded from the database?